lib/test.js

```diff
@@ -262,23 +259,36 @@ class Test {
 				this.endCallbackFinisher = () => {
 					resolve(this.finishPromised());
 				};
+
+				this.finishDueToInactivity = () => {
+					this.saveFirstError(new Error('`t.end()` was never called'));
+					resolve(this.finishPromised());
+				};
 			});
 		}
 
 		if (promise) {
-			return promise.then(
-				() => this.finish(),
-				err => {
-					throwsHelper(err);
-
-					this.saveFirstError(new assert.AssertionError({
-						message: 'Rejected promise returned by test',
-						stack: err instanceof Error && err.stack,
-						values: [formatAssertError.formatWithLabel('Rejection reason:', err)]
-					}));
-					return this.finish();
-				}
-			);
+			return new Promise(resolve => {
+				this.finishDueToInactivity = () => {
+					const err = returnedObservable ?
+						new Error('Observable returned by test never completed') :
+						new Error('Promise returned by test never resolved');
```

Comment:
❤️ This commit is an amazing improvement! I remember trying to achieve something like this in AVA 0.0.2, but I could not get it working. Probably because `beforeExit` didn't exist then.